storage/cloudspanner/log_storage.go

```diff
@@ -379,7 +376,7 @@ func (tx *logTX) StoreSignedLogRoot(ctx context.Context, root *trillian.SignedLo
 			int64(logRoot.TreeSize),
 			logRoot.RootHash,
 			root.LogRootSignature,
-			writeRev,
+			int64(logRoot.Revision),
```

Comment:
Can a caller now write a log root of revision N+M leaving a gap?
Also, would any tiles added in this same transaction be using `writeRev` for _their_ revision, but no be longer guaranteed to match the "corresponding" logRoot revision?

Reply:
Yes, but this is controlled by the sequencer -- as is also the case in the mysql and prod implementations.  Checking this is a business layer concern, not a storage layer one. 

I view this change as correcting what could be a very confusing bug -- roots being stored at a different revision than the one requested.

Reply:
Hmm, I'm not sure I agree about the responsibility - the stacked "swiss cheese" revisions is a storage implementation detail, but with this change I think a caller can corrupt the stored tree by doing:

1. adding some leaves at N, but write a log root at rev N+3
2. add a leaf (which, due to writeRev logic) will be at N+3+1, and store a log root at N+1

etc.

Reply:
Perhaps, but for the sake of consistency I suggest we follow what the [mysql implementation does](https://github.com/google/trillian/blob/master/storage/mysql/log_storage.go#L890).  It's also worth noting that changing the log root will invalidate any log root signatures. See https://github.com/google/trillian/pull/1052

Changing this is an API contract change and should be done consistency across storage implementations and with a corresponding unit test.

Reply:
I take your point on signatures, although that might be a happy side-effect of invalidating an STH which is clearly produced by a logic error :).

I'd argue there has been no explicit API contract, **but** the use of `writeRev` throughout the implementations was intended to enforce the implicit and "reasonably expected" contract that all writes in a given TX would be written at the same revision returned by `WriteRevision()` on the same TX.
FWIW the currently committed implementation here is consistent with the InfraSpanner implementation, BTW.
So I'd say the MySQL behaviour is the odd one out and it's a bug.

As a path forward, how about this for a compromise: this method (and indeed the corresponding other implementations) should return an error if `logRoot.Revision != writeRevision`?

We can add that to the comment for this method on the storage API interface too.